### DISCUSSION TITLE
源码编译报错

### DISCUSSION THREAD

**Discussion by wzg-zhuo** [TuGraph-family/tugraph-db#160] [Q&A]
在centos7编译tugraph源码，cmake成功，但运行make后，报错如下：
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/libgomp.a(icv.o): relocation R_X86_64_32 against hidden symbol `gomp_global_icv' can not be used when making a shared object
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/libgomp.a(parallel.o): relocation R_X86_64_32 against hidden symbol `gomp_global_icv' can not be used when making a shared object
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/libgomp.a(team.o): relocation R_X86_64_TPOFF32 against hidden symbol `gomp_tls_data' can not be used when making a shared object
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/libgomp.a(proc.o): relocation R_X86_64_32 against hidden symbol `gomp_global_icv' can not be used when making a shared object
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/libgomp.a(affinity.o): relocation R_X86_64_32 against `.rodata.str1.8' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/libgomp.a(alloc.o): relocation R_X86_64_32 against `.rodata.str1.8' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/libgomp.a(env.o): relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/libgomp.a(error.o): relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/libgomp.a(task.o): relocation R_X86_64_32S against `.rodata' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/libgomp.a(target.o): relocation R_X86_64_32 against `.text' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/libgomp.a(splay-tree.o): relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/libgomp.a(oacc-init.o): relocation R_X86_64_32S against `.bss' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/libgomp.a(priority_queue.o): relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/libgomp.a(oacc-host.o): relocation R_X86_64_32 against `.data' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(any_lite.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(arena.o): relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(arenastring.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(arenaz_sampler.o): relocation R_X86_64_32 against `.bss' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(extension_set.o): relocation R_X86_64_32 against `.bss' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(generated_enum_util.o): relocation R_X86_64_32 against symbol `_ZN6google8protobuf8internal13DestroyStringEPKv' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(generated_message_tctable_lite.o): relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(generated_message_util.o): relocation R_X86_64_32S against symbol `_ZNSs4_Rep20_S_empty_rep_storageE' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(implicit_weak_message.o): relocation R_X86_64_32 against symbol `_ZN6google8protobuf8internal38implicit_weak_message_default_instanceE' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(inlined_string_field.o): relocation R_X86_64_32S against symbol `_ZNSs4_Rep20_S_empty_rep_storageE' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(coded_stream.o): relocation R_X86_64_32 against `.rodata.str1.8' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(strtod.o): relocation R_X86_64_32 against `.bss' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(zero_copy_stream.o): relocation R_X86_64_32 against `.rodata.str1.8' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(zero_copy_stream_impl.o): relocation R_X86_64_32 against `.rodata.str1.8' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(zero_copy_stream_impl_lite.o): relocation R_X86_64_32S against symbol `_ZTVN6google8protobuf2io19LimitingInputStreamE' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(map.o): relocation R_X86_64_32 against `.bss' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(message_lite.o): relocation R_X86_64_32 against `.rodata.str1.8' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(parse_context.o): relocation R_X86_64_32S against symbol `_ZNSs4_Rep20_S_empty_rep_storageE' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(repeated_field.o): relocation R_X86_64_32 against symbol `_ZTISs' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(repeated_ptr_field.o): relocation R_X86_64_32 against symbol `_ZTIc' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(bytestream.o): relocation R_X86_64_32 against `.rodata.str1.8' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(common.o): relocation R_X86_64_32S against `.rodata' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(int128.o): relocation R_X86_64_32 against `.rodata.str1.8' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(status.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(statusor.o): relocation R_X86_64_32 against `.rodata.str1.8' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(stringpiece.o): relocation R_X86_64_32 against `.rodata.str1.8' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(stringprintf.o): relocation R_X86_64_32S against symbol `_ZNSs4_Rep20_S_empty_rep_storageE' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(structurally_valid.o): relocation R_X86_64_32S against `.rodata' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(strutil.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(time.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(wire_format_lite.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(any.o): relocation R_X86_64_32S against symbol `_ZNSs4_Rep20_S_empty_rep_storageE' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(any.pb.o): relocation R_X86_64_32 against symbol `descriptor_table_google_2fprotobuf_2fany_2eproto' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(api.pb.o): relocation R_X86_64_32 against symbol `descriptor_table_google_2fprotobuf_2fapi_2eproto' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(importer.o): relocation R_X86_64_32S against symbol `_ZTVN6google8protobuf8compiler28SourceTreeDescriptorDatabase24ValidationErrorCollectorE' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(parser.o): relocation R_X86_64_32 against symbol `_ZN6google8protobuf33_MessageOptions_default_instance_E' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(descriptor.o): relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(descriptor.pb.o): relocation R_X86_64_32 against symbol `descriptor_table_google_2fprotobuf_2fdescriptor_2eproto' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(descriptor_database.o): relocation R_X86_64_32S against symbol `_ZTVN6google8protobuf24MergedDescriptorDatabaseE' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(duration.pb.o): relocation R_X86_64_32 against symbol `descriptor_table_google_2fprotobuf_2fduration_2eproto' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(dynamic_message.o): relocation R_X86_64_32S against symbol `_ZTVN6google8protobuf21DynamicMessageFactoryE' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(empty.pb.o): relocation R_X86_64_32 against symbol `descriptor_table_google_2fprotobuf_2fempty_2eproto' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(extension_set_heavy.o): relocation R_X86_64_32S against symbol `_ZN6google8protobuf15FieldDescriptor17kTypeToCppTypeMapE' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(field_mask.pb.o): relocation R_X86_64_32 against symbol `descriptor_table_google_2fprotobuf_2ffield_5fmask_2eproto' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(generated_message_bases.o): relocation R_X86_64_32S against symbol `_ZTVN6google8protobuf8internal14ZeroFieldsBaseE' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(generated_message_reflection.o): relocation R_X86_64_32 against `.rodata.str1.8' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(generated_message_tctable_full.o): relocation R_X86_64_32 against `.bss' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(gzip_stream.o): relocation R_X86_64_32 against `.rodata.str1.8' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(printer.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(tokenizer.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(map_field.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(message.o): relocation R_X86_64_32 against `.text' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(reflection_ops.o): relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(service.o): relocation R_X86_64_32 against `.bss' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(source_context.pb.o): relocation R_X86_64_32 against symbol `descriptor_table_google_2fprotobuf_2fsource_5fcontext_2eproto' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(struct.pb.o): relocation R_X86_64_32 against symbol `descriptor_table_google_2fprotobuf_2fstruct_2eproto' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(substitute.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(text_format.o): relocation R_X86_64_32 against symbol `_ZN6google8protobuf8internal24kTypeGoogleApisComPrefixE' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(timestamp.pb.o): relocation R_X86_64_32 against symbol `descriptor_table_google_2fprotobuf_2ftimestamp_2eproto' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(type.pb.o): relocation R_X86_64_32 against symbol `descriptor_table_google_2fprotobuf_2ftype_2eproto' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(unknown_field_set.o): relocation R_X86_64_32 against `.bss' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(delimited_message_util.o): relocation R_X86_64_32 against `.bss' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(field_comparator.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(field_mask_util.o): relocation R_X86_64_32 against `.rodata.str1.8' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(datapiece.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(default_value_objectwriter.o): relocation R_X86_64_32S against symbol `_ZNSs4_Rep20_S_empty_rep_storageE' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(error_listener.o): relocation R_X86_64_32 against `.bss' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(field_mask_utility.o): relocation R_X86_64_32S against `.rodata.str1.1' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(json_escaping.o): relocation R_X86_64_32S against `.rodata' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(json_objectwriter.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(json_stream_parser.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(object_writer.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(proto_writer.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(protostream_objectsource.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(protostream_objectwriter.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(type_info.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(utility.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(json_util.o): relocation R_X86_64_32S against symbol `_ZTVN6google8protobuf4util8internal22ZeroCopyStreamByteSinkE' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(message_differencer.o): relocation R_X86_64_32 against `.rodata.str1.8' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(time_util.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(type_resolver_util.o): relocation R_X86_64_32S against `.rodata' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(wire_format.o): relocation R_X86_64_32 against symbol `__pthread_key_create@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: //usr/local/lib/libprotobuf.a(wrappers.pb.o): relocation R_X86_64_32 against symbol `descriptor_table_google_2fprotobuf_2fwrappers_2eproto' can not be used when making a shared object; recompile with -fPIC
/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld: final link failed: Nonrepresentable section on output
collect2: error: ld returned 1 exit status
make[2]: *** [output/liblgraph.so] Error 1
make[1]: *** [src/CMakeFiles/lgraph.dir/all] Error 2
make: *** [all] Error 2

**Comment by jiazhenjiang** [comment]
Please try to run the following command：
sed -i '3 s/-lgomp/-l:libgomp.a/' /usr/local/lib64/libgomp.spec

**Comment by wzg-zhuo** [comment]
@jiazhenjiang  
not work

**Comment by jiazhenjiang** [comment]
You can recompile the gcc toolchain, adding the -fPIC parameter at compile time, like this
wget http://ftp.gnu.org/gnu/gcc/gcc-8.4.0/gcc-8.4.0.tar.gz \
    && tar -xvf gcc-8.4.0.tar.gz && cd gcc-8.4.0 \
    && ./contrib/download_prerequisites && mkdir build && cd build \
    && ../configure CFLAGS=-fPIC CXXFLAGS=-fPIC -enable-checking=release -enable-languages=c,c++ -disable-multilib \
    && make -j4 && make install \
    && cp  ./x86_64-*-linux-gnu/libstdc++-v3/src/.libs/libstdc++.so.6.0.25 /usr/lib64/ \
    && cd /usr/lib64 && ln -sf libstdc++.so.6.0.25 libstdc++.so.6 \
    && rm -rf /gcc-* && ln -s /usr/local/bin/gcc /usr/local/bin/cc

**Comment by yxdai-nju** [comment]
I encountered a similar problem when using the `tugraph/tugraph-compile-centos8:1.3.2` Docker image on my M2-chip MacBook to build tugraph-db:

```
[root@33f44bc5d57e build]# make -f src/CMakeFiles/lgraph.dir/build.make src/CMakeFiles/lgraph.dir/build
[  0%] Linking CXX shared library ../output/liblgraph.so
/usr/bin/ld: /usr/lib/gcc/x86_64-redhat-linux/8/libgomp.a(team.o): relocation R_X86_64_TPOFF32 against hidden symbol `gomp_tls_data' can not be used when making a shared object
/usr/bin/ld: /usr/lib/gcc/x86_64-redhat-linux/8/libgomp.a(oacc-init.o): relocation R_X86_64_TPOFF32 against hidden symbol `goacc_tls_data' can not be used when making a shared object
/usr/bin/ld: final link failed: Nonrepresentable section on output
```

~~Surprisingly, the issue seems to be related to Rosetta, which emulates x86_64/amd64 on Apple Silicon chips.~~

~~Try turning off "Use Rosetta for x86_64/amd64 emulation on Apple Silicon" in your Docker Desktop settings and attempt to build again. This might resolve the problem.~~

Edit on July 21, 2024: I apologize for my earlier misunderstanding. The issue is not related to Rosetta. 

The ".../gcc/x86_64-redhat-linux/7/..." paths in your error messages indicate you're using the system default GCC toolchain. Instead, you should use the GCC toolchain located at /usr/local/bin/gcc and /usr/local/bin/g++, which was compiled with the `-fPIC` flag as @jiazhenjiang suggested.

To resolve this, update your compiler configuration with something like `CC="/usr/local/bin/gcc"` and `CXX="/usr/local/bin/g++"`. This should address the relocation errors you're encountering during the build process.
